feat(snapshot): browse offline endpoint based on snapshot

xAt0mZ opened the pull request:
Closes #2221 .

TODO list, updated during development progress:

_Persisted data_
- [ ] Containers
- [ ] Volumes
- [ ] Networks
- [ ] Images
- [ ] Stacks
- [ ] Engine

_Compatible views_
- [ ] Dashboard
- [ ] Container list
- [ ] Volume list
- [ ] Network list
- [ ] Image list
- [ ] Engine

xAt0mZ commented:
@deviantony updated, should be working now.
I can't test Azure endpoints, but should be working aswell (swarm blocked, standalone working offline)

deviantony commented:
Snapshots are not enabled for Azure environments, this feature should only be related to Docker environments.

xAt0mZ commented:
> I've started a Portainer instance with this implementation using existing Portainer data and was not able to explore a down endpoint:

May I have some browser network debug logs ? I'm not able to reproduce this behaviour ...

> And it seems that if I access it now and go back to the home view, it stays as 'Down'. So not sure about how to reproduce this one.

This is the desync between the snapshot data sent by the API and the real status of the endpoint. It has always been there since we have this snapshot feature.
We can only detect and update the endpoint status while trying to access it (or access a view of it).
We can't modify the snapshot value, as it's updated by the API with `/endpoints` route every time we come back to the `Home` view.

> What happens when the --no-snapshot flag as been set or if there is no snapshot available for the endpoint when accessing the offline views?

As long as you need to select the endpoint in the `Home` view, if there is no snapshot available, your attempt should be rejected. I forgot to do it [there](https://github.com/portainer/portainer/pull/2253/commits/9821bf42b36a60379580b6a5d35e0e9ebbb6e679#diff-cd0ce8b7e362e6513866c44176d03208R14)

xAt0mZ commented:
@deviantony updated. Should be working as expected now.
Fixed another bug where trying to access a down endpoint and failing changed the endpoint ID and publicURL in the global state to the ones of the down endpoint.


Small things:
* sometimes loading bar doesn't show up as we chain requests (`cfpLoadingbar` hasn't time to reset and restart). I modified the Threshold, but it may bug for some clients, depending on the machine performances and overall usage.
* still no "greyed out" down endpoints, as it will require us to modify the state, ping & evaluate if the endpoint is allowed to be browsed offline, and disable it if necessary.
Home view may have a way too long loading time if a client has a lot of down endpoints to check (max 10s * down endpoints count).

deviantony commented:
@xAt0mZ I'd really like the status update upon accessing the dashboard of any endpoint.

Also, could you rebased the PR now?

deviantony commented:
@xAt0mZ still need a rebase.

xAt0mZ commented:
@deviantony updated !

xAt0mZ commented:
@deviantony updated

deviantony commented on api/portainer.go:
```diff
@@ -245,17 +245,28 @@ type (
+		SnapshotRaw           SnapshotRaw `json:"SnapshotRaw"`
+	}
+
+	// SnapshotRaw represents all the information related to a snapshot
```
Description should be `represents all the information related to a snapshot as returned by the Docker API`.

deviantony commented on app/docker/components/datatables/containers-datatable/containersDatatable.html:
```diff
@@ -99,7 +99,7 @@
         </div>
       </div>
-      <containers-datatable-actions
+      <containers-datatable-actions ng-if="$ctrl.state.endpointStatus === 1"
```
I think that any part of a component that should be hidden related to the status of an endpoint should be referring to a binding such as `ng-if="!$ctrl.offlineMode"` instead of relying on the `endpointStatus` being retrieved inside each component (via `EndpointProvider`, see my comment below for more details).

deviantony commented on app/docker/components/datatables/containers-datatable/containersDatatableController.js:
```diff
@@ -13,7 +13,8 @@ function (PaginationService, DatatableService, EndpointProvider) {
     noPausedItemsSelected: true,
-    publicURL: EndpointProvider.endpointPublicURL()
+    publicURL: EndpointProvider.endpointPublicURL(),
+    endpointStatus: EndpointProvider.endpointStatus()
```
Related to my comment above, instead of tying this component (and the others that requires this information) to the `EndpointProvider`, I believe that we should relay on a binding (such as `offlineMode` as specified above) to inject this property in the component at a higher level.

This will probably makes things easier for us if we decide that we don't way to relay on `EndpointProvider` anymore at some point in time.

Should be applied for all components actually using this logic.

deviantony commented on app/docker/components/datatables/networks-datatable/networksDatatable.js:
```diff
@@ -10,6 +10,7 @@ angular.module('portainer.docker').component('networksDatatable', {
     showHostColumn: '<',
-    removeAction: '<'
+    removeAction: '<',
+    endpointStatus: '<'
```
Simply pass the newly created binding `offlineMode` here instead of passing the status.

deviantony commented on app/docker/components/dockerSidebarContent/dockerSidebarContent.html:
```diff
@@ -1,7 +1,7 @@
   <a ui-sref="docker.dashboard" ui-sref-active="active">Dashboard <span class="menu-icon fa fa-tachometer-alt fa-fw"></span></a>
 </li>
-<li class="sidebar-list">
+<li class="sidebar-list" ng-if="$ctrl.endpointStatus === 1">
```
Same here, use `$ctrl.offlineMode`.

deviantony commented on app/docker/rest/system.js:
```diff
@@ -1,16 +1,16 @@
+      method: 'GET', params: { action: 'info' }, timeout: 10000, interceptor: InfoInterceptor
     },
-    version: { method: 'GET', params: { action: 'version' }, ignoreLoadingBar: true, timeout: 4500 },
+    version: { method: 'GET', params: { action: 'version' }, timeout: 4500, interceptor: VersionInterceptor },
```
Any particular reason to remove `ignoreLoadingBar: true` here?

xAt0mZ replied:
`Home` view does 3 API calls before going to `Dashboard` : `_ping`, `version` and `info`.
If we ignore the loading bar for `version`, you will have it for `ping`, then nothing for `version`, then have it back for `info`.
When the user is waiting for the `version` call to come back when browsing offline without loading bar, it's not UX friendly, as he will think the app is not working (nothing indicates the app is waiting for a response).

deviantony commented on app/portainer/components/information-panel/information-panel.js:
```diff
@@ -2,7 +2,8 @@ angular.module('portainer.app').component('informationPanel', {
     titleText: '@',
-    dismissAction: '&'
+    dismissAction: '&',
+    cantDismiss: '<'
```
You can remove the `cantDismiss` binding, see my comment below.

deviantony commented on app/portainer/components/information-panel/informationPanel.html:
```diff
@@ -6,7 +6,7 @@
             {{ $ctrl.titleText }}
           </span>
-          <span class="small" style="float: right;">
+          <span class="small" style="float: right;" ng-if="!$ctrl.cantDismiss">
```
You can simply relay on the fact that the `dismissAction` binding was specified here: `ng-if="$ctrl.dismissAction"`

xAt0mZ replied:
Changing to `ng-if="$ctrl.dismissAction"` doesn't hide the button even if the `dismissAction` binding isn't provided.

When using the `&` binding, angular wraps the function in order to keep references to the original $scope of the passed method, even if is not defined.
Function is :
```
ƒ (locals) {
   return parentGet(scope, locals);
}
```

deviantony commented on app/docker/views/containers/containers.html:
```diff
@@ -6,7 +6,14 @@
   <rd-header-content>Containers</rd-header-content>
 </rd-header>
-
+<information-panel title-text="Offline mode" cant-dismiss="true" ng-if="endpointStatus !== 1">
```
Please remove any `cant-dismiss` attribute in other views as well (see my comment regarding information-panel.js).

deviantony commented on app/portainer/services/api/endpointService.js:
```diff
@@ -16,6 +16,10 @@ function EndpointServiceFactory($q, Endpoints, FileUploadService) {
     return Endpoints.snapshot({}, {}).$promise;
   };
 
+  service.checkEndpointStatus = function(endpointId) {
```
Seems redundant with the usage of `SystemService.ping`. What's the point of wrapping this function? Can't you use `SystemService.ping()` instead of `EndpointService.checkEndpointStatus()` ?

deviantony commented on app/portainer/services/endpointProvider.js:
```diff
@@ -37,5 +52,33 @@ angular.module('portainer.app')
+    LocalStorage.storeEndpointStatus(status);
+  };
+
+  service.currentEndpoint = function() {
```
This whole section seems like a big workaround for me but it will be ok for now. Ultimately we need to think about a way to store the entire endpoint object instead of storing properties like id and public URL.

chiptus replied:
Maybe use JSON.parse/stringify?

deviantony commented on app/portainer/views/home/homeController.js:
```diff
@@ -1,95 +1,132 @@
+      $scope.goToDashboard = function (endpoint) {
+        var status = 1;
+
+        EndpointService.checkEndpointStatus(endpoint.Id)
```
See my comment above about using SystemService.ping here instead of this function that is simply wrapping it.

deviantony commented on app/portainer/views/home/homeController.js:
```diff
@@ -1,95 +1,132 @@
+          }).catch(function error() {
+            status = 2;
+          }).finally(function () {
+            if (endpoint.Status !== status) {
```
Try to refactor this block to smaller functions, it's quite hard to read.

chiptus commented on app/docker/interceptors/containersInterceptor.js:
```diff
@@ -0,0 +1,17 @@
+angular.module('portainer.app')
```
I'm not sure how, but the interceptors code is repeating itself. Maybe there's a way to make it DRY?

chiptus commented on app/docker/views/networks/networks.html:
```diff
@@ -6,7 +6,14 @@
   <rd-header-content>Networks</rd-header-content>
 </rd-header>
-
+<information-panel title-text="Offline mode" cant-dismiss="true" ng-if="endpointStatus !== 1">
```
since it's the same information-panel. Maybe we can make a component out of this?

deviantony replied:
Agree, something like `information-panel-offline`.

chiptus commented on app/portainer/views/home/homeController.js:
```diff
@@ -1,95 +1,132 @@
-    });
-  }
+      function switchToAzureEndpoint(endpoint) {
+        EndpointProvider.setEndpointID(endpoint.Id);
```
I think endpoint provider should have `.setEnpoint` function, I guess this happens anytime on changing endpoint

deviantony replied:
IMO should be part of a specific refactor that only targets the state manager. Related to endpoint provider refactor as well.

chiptus commented on app/portainer/views/sidebar/sidebarController.js:
```diff
@@ -1,36 +1,37 @@
 angular.module('portainer.app')
```
Anything was changed here beside code style? If it's only style I don't think it should be in this PR

deviantony replied:
Agree, please avoid the use of formatting tools to make it easier to review.

chiptus commented:
I'm not sure if it is suppose to be like that. But when clicking an endpoint it doesn't show any loading bar and screen doesn't change. If the endpoint is offline than ping takes time (a few seconds) in which the user just waits. In the first times I thought my click didn't register so I clicked a few times.

chiptus commented:
When following this:

> Deploy a dind container that will be used as an endpoint: docker run -d --privileged -p 3000:2375 --name endpoint01 docker:dind
> 
> Create a new Docker endpoint, use URL: myIP:3000.
> 
> Explore the endpoint via the home view, looks fine. Create a container, go back to home, and click on refresh to trigger a snapshot.
> 
> Go to your local endpoint, and stop the endpoint01 container.

I can open the snapshot of the endpoint, but my created container is not shown in the containers list.
After starting the `endpoint01` container and refreshing the endpoints list (this makes a new snapshot?) it works, even after stopping the container.
I guess that we should update the snapshot when creating new resources

deviantony commented on api/docker/snapshot.go:
```diff
@@ -152,5 +165,24 @@ func snapshotVolumes(snapshot *portainer.Snapshot, cli *client.Client) error {
 	}
 
 	snapshot.VolumeCount = len(volumes.Volumes)
+	snapshot.SnapshotRaw.Volumes = volumes
```
No, `SnapshotRaw` has to respect the format of the Docker API response.

deviantony commented:
> I guess that we should update the snapshot when creating new resources

Regarding @chiptus remark, no, for now we'll only rely on the background snapshot feature to refresh the snapshot data. Could be an enhancement later tho.

xAt0mZ commented:
@chiptus 
> I'm not sure if it is suppose to be like that. But when clicking an endpoint it doesn't show any loading bar and screen doesn't change. If the endpoint is offline than ping takes time (a few seconds) in which the user just waits. In the first times I thought my click didn't register so I clicked a few times.

We don't ignore loadingBar on ping request, so it should show up. Actually I faced this behaviour (loadingBar not showing) when been redirected from home to the dashboard. That's why I changed the `cfpLoadingBarProvider.latencyThreshold`.

The cpfLoadingbar doesn't really like when we are chaining API calls ... Issues and potential fixes [here](https://github.com/chieffancypants/angular-loading-bar/issues/359) and [here](https://github.com/chieffancypants/angular-loading-bar/issues/377)
